File: scripts/build-video-og.js
/**
 * build-video-og.js — OG del video (video_og.jpg, 1200×630) para el paso 🎬 del pipeline (tras resolver videos).
 * Lee /tmp/triggui-book.json (slug) + contenido.json (con _video). Misma plantilla que scripts/build-video-og.py.
 * Nunca rompe el pipeline: cualquier fallo se reporta y sale con 0.
 */
import fs from "node:fs/promises";
import path from "node:path";
import { chromium } from "playwright";
const esc = (s) => String(s ?? "").replace(/[&<>"']/g, (c) => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" }[c]));
const LOGO = "https://raw.githubusercontent.com/badirnakid/triggui-app/main/public/trigguiletrasblanco2.png";
async function thumb(id) {
  for (const q of ["maxresdefault", "hqdefault"]) {
    const u = `https://i.ytimg.com/vi/${id}/${q}.jpg`;
    try { const r = await fetch(u, { method: "HEAD" }); if (r.ok && Number(r.headers.get("content-length") || 0) > 2000) return u; } catch {}
  }
  return `https://i.ytimg.com/vi/${id}/hqdefault.jpg`;
}
try {
  const meta = JSON.parse(await fs.readFile("/tmp/triggui-book.json", "utf8"));
  const cat = process.argv[2] || "contenido.json";
  const libros = JSON.parse(await fs.readFile(cat, "utf8")).libros || [];
  const libro = libros.find((l) => l._slug === meta.slug) || libros.find((l) => l.titulo === meta.titulo);
  if (!libro) { console.log(`ℹ️  sin libro para ${meta.slug}`); process.exit(0); }
  const vc = ((libro._video || {}).candidatos || []).filter((x) => x && x.id).slice(0, 3);
  if (!vc.length) { console.log("ℹ️  sin video"); process.exit(0); }
  const outBase = process.env.TRIGGUI_OUT_BASE || "public/t";
  const outDir = path.join(outBase, meta.slug); await fs.mkdir(outDir, { recursive: true });
  const cols = Array.isArray(libro.colores) ? libro.colores : []; const A = cols[0] || "#E8A838", B = cols[1] || A;
  const tpl = await fs.readFile("scripts/templates/og-video.html", "utf8");
  const browser = await chromium.launch({ headless: true }); const page = await browser.newPage({ viewport: { width: 1200, height: 630 } });
  for (let i = 0; i < vc.length; i++) {
    const v = vc[i]; const tit = String(v.titulo || "");
    const html = tpl.replace("{{ACCENT}}", esc(A)).replace("{{ACCENT2}}", esc(B)).replace("{{THUMB}}", await thumb(v.id))
      .replace("{{CLASE}}", tit.length > 60 ? " larga" : "").replace("{{TITULO}}", esc(tit.slice(0, 140))).replace("{{CANAL}}", esc(v.canal || ""))
      .replace("{{LOGO}}", LOGO).replace("{{LIBRO}}", esc(libro.titulo || ""));
    await page.setContent(html, { waitUntil: "networkidle" }); await page.waitForTimeout(250);
    const out = path.join(outDir, i === 0 ? "video_og.jpg" : `video_og_${i}.jpg`);
    await page.screenshot({ path: out, type: "jpeg", quality: 86, clip: { x: 0, y: 0, width: 1200, height: 630 } });
    console.log(`   🎬 OG video ${i}: ${out}`);
  }
  await browser.close();
} catch (e) { console.log(`   ⚠️ OG video omitido: ${e && e.message}`); }

File: scripts/build-video-og.py
#!/usr/bin/env python3
# 🎬 build-video-og.py — OG del video (1200×630): miniatura de YouTube + título + canal, sobre el degradado de la edición.
# Uso: python3 scripts/build-video-og.py <catalogo.json> <slug> <out_dir>
import sys, json, html, asyncio, urllib.request
from playwright.async_api import async_playwright
LOGO = "https://raw.githubusercontent.com/badirnakid/triggui-app/main/public/trigguiletrasblanco2.png"
def esc(s): return html.escape(str(s or ""), quote=True)
def thumb(vid):
    for q in ("maxresdefault", "hqdefault"):
        u = f"https://i.ytimg.com/vi/{vid}/{q}.jpg"
        try:
            r = urllib.request.urlopen(urllib.request.Request(u, headers={"User-Agent": "Mozilla/5.0"}), timeout=8)
            if r.status == 200 and int(r.headers.get("Content-Length") or 0) > 2000: return u
        except Exception: pass
    return f"https://i.ytimg.com/vi/{vid}/hqdefault.jpg"
async def render(libro, out_path):
    """out_path = <out_dir>/video_og.jpg; escribe también video_og_1.jpg y video_og_2.jpg (uno por candidato)."""
    vc = [x for x in ((libro.get("_video") or {}).get("candidatos") or []) if x.get("id")][:3]
    if not vc: return False
    col = libro.get("colores") or []; a = col[0] if col else "#E8A838"; b = col[1] if len(col) > 1 else a
    base = open("scripts/templates/og-video.html", encoding="utf-8").read()
    async with async_playwright() as p:
        br = await p.chromium.launch(); pg = await br.new_page(viewport={"width": 1200, "height": 630})
        for i, v in enumerate(vc):
            tit = v.get("titulo", "")
            t = (base.replace("{{ACCENT}}", esc(a)).replace("{{ACCENT2}}", esc(b)).replace("{{THUMB}}", thumb(v["id"]))
                  .replace("{{CLASE}}", " larga" if len(tit) > 60 else "").replace("{{TITULO}}", esc(tit[:140])).replace("{{CANAL}}", esc(v.get("canal", "")))
                  .replace("{{LOGO}}", LOGO).replace("{{LIBRO}}", esc(libro.get("titulo", ""))))
            await pg.set_content(t, wait_until="networkidle"); await pg.wait_for_timeout(250)
            dest = out_path if i == 0 else out_path.replace("video_og.jpg", f"video_og_{i}.jpg")
            await pg.screenshot(path=dest, type="jpeg", quality=86, clip={"x": 0, "y": 0, "width": 1200, "height": 630})
        await br.close()
    return True
if __name__ == "__main__":
    cat, slug, out_dir = sys.argv[1], sys.argv[2], sys.argv[3]
    libro = next((l for l in json.load(open(cat, encoding="utf-8"))["libros"] if l.get("_slug") == slug), None)
    if not libro: print("sin libro para", slug); sys.exit(2)
    print(("✅ " if asyncio.run(render(libro, f"{out_dir}/video_og.jpg")) else "— sin video ") + slug)
